sdk/core/azure-core/src/main/java/com/azure/core/util/ClientOptions.java

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import com.azure.core.annotation.Fluent;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.core.http.policy.UserAgentPolicy;
+
+import java.util.ArrayList;
+import java.util.Objects;
+
+/**
+ * This class represents various options to be set on the client. It is set on the builder at the time of building the
+ * client.
+ * <p>
+ * The {@link Header} could be set using {@link ClientOptions#setHeaders(Iterable) setHeaders}. The {@link Header} will
+ * be applied on the request being sent to Azure Service.
+ * <p>
+ * The {@code applicationId} could be set using {@link ClientOptions#setApplicationId(String) setApplicationId} which
+ * is used for setting {@code applicationId} in the {@link UserAgentPolicy}.
+ */
+@Fluent
+public final class ClientOptions {
+    private static final int MAX_APPLICATION_ID_LENGTH = 24;
+    private final ClientLogger logger = new ClientLogger(ClientOptions.class);
+    private Iterable<Header> headers;
```

Comment:
Any reason to require the user to set the entire `Iterable` rather than offering a `setHeaders(Header...)`, or `setHeaders(Iterable<Header>)`, and `addHeader(Header)` operators? I feel this will lead to a lot of situations where the user will need to use `Collections.singletonList()`.

Reply:
I had this specific discussion with @srnagar .  Earlier in the PR I did had `addHeader(name, value), remove(name), getValue(name)` API but the argument against this was 1. This class is not equivalent of HttpHeaders  2. This is not primary way to add the Headers . We do not want every time user create a client, they have to set a Header here.  It is kind of hidden, user use them in special cases.  So need to add these additional API is not so strong. But in future if we see usage pattern like this, we can add additional API.